src/strbuf.c

```diff
@@ -0,0 +1,483 @@
+        strbuf->magic = STRBUF_MAGIC;
+        strbuf->maxsize = size;
+        strbuf->size = 0;
+        strbuf->data[0] = EOS;
+    }
+
+    return strbuf;
+}
+
+void Strbuf_Free(Strbuf *strbuf)
+{
+    Strbuf_Check(strbuf);
+
+    strbuf->magic = STRBUF_MAGIC + 1;
+    Heap_FreeToHeap(strbuf);
+}
+
+void Strbuf_Clear(Strbuf *strbuf)
+{
+    Strbuf_Check(strbuf);
+
+    strbuf->size = 0;
+    strbuf->data[0] = EOS;
+}
+
+void Strbuf_Copy(Strbuf *dst, const Strbuf *src)
+{
+    Strbuf_Check(dst);
+    Strbuf_Check(src);
+
+    if (dst->maxsize > src->size) {
+        memcpy(dst->data, src->data, (src->size + 1) * sizeof(Charcode));
+        dst->size = src->size;
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+Strbuf* Strbuf_Clone(const Strbuf *src, u32 heapID)
+{
+    Strbuf_Check(src);
+
+    Strbuf *strbuf = Strbuf_Init(src->size + 1, heapID);
+
+    if (strbuf) {
+        Strbuf_Copy(strbuf, src);
+    }
+
+    return strbuf;
+}
+
+void Strbuf_FormatInt(Strbuf *dst, int num, u32 maxDigits, PaddingMode paddingMode, BOOL whichCharset)
+{
+    static const u32 sPowersOfTen[] = {
+        1,
+        10,
+        100,
+        1000,
+        10000,
+        100000,
+        1000000,
+        10000000,
+        100000000,
+        1000000000
+    };
+    static const Charcode sDigits_JP[] = {
+        JP_0,
+        JP_1,
+        JP_2,
+        JP_3,
+        JP_4,
+        JP_5,
+        JP_6,
+        JP_7,
+        JP_8,
+        JP_9,
+    };
+    static const Charcode sDigits_EN[] = {
+        EN_0,
+        EN_1,
+        EN_2,
+        EN_3,
+        EN_4,
+        EN_5,
+        EN_6,
+        EN_7,
+        EN_8,
+        EN_9,
+    };
+
+    Strbuf_Check(dst);
+
+    BOOL negative = (num < 0);
+
+    if (dst->maxsize > (maxDigits + negative)) {
+        const Charcode *digitSet = (whichCharset == 0) ? sDigits_JP : sDigits_EN;
+
+        Strbuf_Clear(dst);
+
+        if (negative) {
+            num *= -1;
+            dst->data[dst->size++] = (whichCharset == 0) ? JP_MINUS : EN_MINUS;
+        }
+
+        u32 div = sPowersOfTen[maxDigits - 1];
+        while (div != 0) {
+            u16 digit = (u16)(num / div);
+            num -= div * digit;
+
+            if (paddingMode == PADDING_MODE_ZEROES) {
+                dst->data[dst->size++] = (digit < 10) ? digitSet[digit] : JP_QUESTION;
+            // If we hit a non-zero digit, flip the padding mode off
+            } else if (digit != 0 || div == 1) {
+                paddingMode = PADDING_MODE_ZEROES;
+                dst->data[dst->size++] = (digit < 10) ? digitSet[digit] : JP_QUESTION;
+            } else if (paddingMode == PADDING_MODE_SPACES) {
+                dst->data[dst->size++] = (whichCharset == 0) ? JP_SPACE : NUM_SPACE;
+            }
+
+            div /= 10;
+        }
+
+        dst->data[dst->size] = EOS;
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+void Strbuf_FormatU64(Strbuf *dst, u64 num, u32 maxDigits, PaddingMode paddingMode, int whichCharset)
```

Comment:
Functionally, they are the same; the awkward thing is that the checks are against 0, so it can't just be something like:

```c
const charcode_t *digitSet = (useJPCharset) ? sDigits_JP : sDigits_EN;
```

It's also kinda weird to me to classify this as a `BOOL` in that sense; what does `Strbuf_FormatU64(..., FALSE)` even mean?

Maybe it's better to take this back to being an enum.

Reply:
Yeah, I'm flipping this back to an enum. It matches with it, and I think it reads better.